core/server/worker/src/main/java/alluxio/worker/block/UnderFileSystemBlockReader.java

```diff
@@ -0,0 +1,324 @@
+
+  /** The initial size of the block allocated in Alluxio storage when the block is cached. */
+  private final long mInitialBlockSize;
+  /** The block metadata for the UFS block. */
+  private final UnderFileSystemBlockMeta mBlockMeta;
+  /** If set, do not cache the block. */
+  private final boolean mNoCache;
+  /** The Alluxio block store. It is used to interact with Alluxio. */
+  private final BlockStore mAlluxioBlockStore;
+
+  /** The input stream to read from UFS. */
+  private InputStream mUnderFileSystemInputStream;
+  /** The block writer to write the block to Alluxio. */
+  private LocalFileBlockWriter mBlockWriter;
+  /** If set, the reader is closed and should not be used afterwards. */
+  private boolean mClosed;
+
+  /**
+   * The position of mUnderFileSystemInputStream (if not null) is blockStart + mInStreamPos.
+   * When mUnderFileSystemInputStream is not set, this is set to -1 (an invalid state) to indicate
+   * that the input stream is not initialized.
+   */
+  private long mInStreamPos;
+
+  /**
+   * Creates an instance of {@link UnderFileSystemBlockReader} and initializes it with a reading
+   * offset.
+   *
+   * @param blockMeta the block meta
+   * @param offset the position within the block to start the read
+   * @param noCache do not cache the block if set
+   * @param alluxioBlockStore the Alluxio block store
+   * @return the block reader
+   * @throws BlockDoesNotExistException if the UFS block does not exist in the UFS block store
+   * @throws IOException if an I/O related error occur
+   */
+  public static UnderFileSystemBlockReader create(UnderFileSystemBlockMeta blockMeta, long offset,
+      boolean noCache, BlockStore alluxioBlockStore)
+      throws BlockDoesNotExistException, IOException {
+    UnderFileSystemBlockReader ufsBlockReader =
+        new UnderFileSystemBlockReader(blockMeta, noCache, alluxioBlockStore);
+    ufsBlockReader.init(offset);
+    return ufsBlockReader;
+  }
+
+  /**
+   * Creates an instance of {@link UnderFileSystemBlockReader}.
+   *
+   * @param blockMeta the block meta
+   * @param noCache do not cache the block
+   * @param alluxioBlockStore the Alluxio block store
+   */
+  private UnderFileSystemBlockReader(UnderFileSystemBlockMeta blockMeta, boolean noCache,
+      BlockStore alluxioBlockStore) {
+    mInitialBlockSize = Configuration.getBytes(PropertyKey.WORKER_FILE_BUFFER_SIZE);
+    mBlockMeta = blockMeta;
+    mAlluxioBlockStore = alluxioBlockStore;
+    mNoCache = noCache;
+    mInStreamPos = -1;
+  }
+
+  /**
+   * Initializes the reader. This is only called in the factory method.
+   *
+   * @param offset the position within the block to start the read
+   * @throws BlockDoesNotExistException if the UFS block does not exist in the UFS block store
+   * @throws IOException if an I/O related error occur
+   */
+  private void init(long offset) throws BlockDoesNotExistException, IOException {
```

Comment:
@peisun1115 I think @calvinjia means we could probably inline `init()` in factory method rather than here.
I have no preference on this. leave this up to you